app/services/data_migrations/delete_work_history_audits.rb

```diff
@@ -0,0 +1,24 @@
+module DataMigrations
+  class DeleteWorkHistoryAudits
+    TIMESTAMP = 20240917154444
+    MANUAL_RUN = true
+    BATCH_SIZE = 1500
+
+    def change
+      relation.in_batches(of: BATCH_SIZE) do |batch|
+        DeleteAuditsWorker.perform_async(batch.ids)
+      end
```

Comment:
I will put here my opinion but feel free to choose to go forward or not with my views. It is just me being over cautious.

When you deleted locally the million of records how long did it take?
While it was deleted did you try to use local apply and see if the DB wasn't busy?

My main concern is deleting the **3 million at one go**. 

**I'd recommend to use Batch Delivery to stagger for over a hour (or more?).** Process deletions gradually in time would be a safer option.

The reason is deleting millions of records at one go (the workers processing all at the same time) can still result in locking issues on the audited_audit table. This could block other queries that need to access these records while the deletion is in progress.

Also another thing: how much DB disk space it took while the deletion was happening? 
PostgreSQL uses MVCC (Multi-Version Concurrency Control) - this mean when a row is deleted, it isn't immediately removed from disk. Instead, it's marked as "dead" and made invisible to future transactions. The actual space is reclaimed later through a process called vacuuming. 
So disk space could sky rocket if we delete million of rows. 
And similar to table bloat, indexes will accumulate dead entries, which also need to be cleaned up by vacuuming.
If we have enough DB disk space, we shouldn't worried but would be nice **to know how much DB disk usage percentage increases to avoid surprises.**

Last thing: probably is in your radar, but let's run this large delete operations during off-peak times to minimize any impact.

Reply:
I did use the batch delivery and it looked like it did a count(to spread them out I think) on all the records that we want to target, 3 million and it loaded all columns of the audit table. The PR only loads the id and it doesn't do the count

So I thought these extra steps were not necessary.

Don't know how much DB disk space we have, I'll try to find out 

@dcyoung-dev what do you think? Should we spread the workers a bit?